write-fonts/src/tables/cmap.rs

```diff
@@ -237,6 +236,187 @@ impl Cmap {
+    id_delta: Option<i32>,
+}
+
+impl Format4Segment {
+    fn len(&self) -> usize {
+        self.end_ix - self.start_ix + 1
+    }
+
+    // cost in bytes of this segment
+    fn cost(&self) -> usize {
+        const BASE_COST: usize = 8; // 4 u16s for a new segment
+        let glyph_id_cost = self
+            .id_delta
+            .is_none()
+            .then(|| self.len() * u16::RAW_BYTE_LEN)
+            .unwrap_or(0);
+        BASE_COST + glyph_id_cost
+    }
+
+    fn combine(&self, other: Format4Segment) -> Format4Segment {
+        assert_eq!(other.start_ix, self.end_ix + 1);
+        Format4Segment {
+            start_ix: self.start_ix,
+            end_ix: other.end_ix,
+            id_delta: None,
+        }
+    }
+}
+
+impl<'a> Format4Ranges<'a> {
+    fn new(mappings: &'a [(char, GlyphId)]) -> Self {
+        // ignore chars above BMP:
+        let mappings = mappings
+            .iter()
+            .position(|(c, _)| u16::try_from(*c as u32).is_err())
+            .map(|bad_idx| &mappings[..bad_idx])
+            .unwrap_or(mappings);
+        Self {
+            mappings,
+            seg_start: 0,
+            currently_contiguous: false,
+        }
+    }
+
+    /// a convenience method called from our iter in the various cases where
+    /// we emit a segment.
+    ///
+    /// a 'seg_len' of 0 means start == end, e.g. a segment of one glyph.
+    fn make_segment(&mut self, seg_len: usize) -> Format4Segment {
+        // if start == end, we should always use a delta.
+        let use_delta = self.currently_contiguous || seg_len == 0;
+        let result = Format4Segment {
+            start_ix: self.seg_start,
+            end_ix: self.seg_start + seg_len,
+            id_delta: use_delta
+                .then_some(
+                    self.mappings
+                        .get(self.seg_start)
+                        .map(|(cp, gid)| gid.to_u32() as i32 - *cp as u32 as i32),
+                )
+                .flatten(),
+        };
+        self.seg_start += seg_len + 1;
+        self.currently_contiguous = false;
+        result
+    }
+}
+
+// this iterator creates segments greedily, that is it will create a new segment
+// at every opportunity. These are then recombined by the caller to generate
+// the most efficient overall sequence of segments.
+impl<'a> Iterator for Format4Ranges<'a> {
+    type Item = Format4Segment;
+    fn next(&mut self) -> Option<Format4Segment> {
+        if self.seg_start == self.mappings.len() {
+            return None;
+        }
+
+        let Some(((mut prev_cp, mut prev_gid), rest)) =
+            self.mappings[self.seg_start..].split_first()
+        else {
+            // if this is the last element, make a final segment
+            return Some(self.make_segment(0));
+        };
+
+        for (i, (cp, gid)) in rest.iter().enumerate() {
+            // first: all segments must be a contiguous range of codepoints
+            if *cp as u32 - prev_cp as u32 > 1 {
```

Comment:
The check is throwing me, can' t we just check `if cp != prev_cp + 1`, e.g. more directly say "if cp doesn't continue the run"